### PR TITLE
Thea: Fisherman and Batch Signing

### PR DESCRIPTION
**Pull request by Gauthamastro**
## Describe your changes

This PR raises the final changes to core Thea logic by adding Fisherman model for deposits and batch signing for withdrawals

### CODE REVIEW

**Comment by Gauthamastro** [comment]
@zktony Please add tests for this PR. Let me know if you need any help

File: pallets/thea-executor/src/lib.rs [hunk 1]
```diff
@@ -402,6 +403,30 @@ pub mod pallet {
 			Self::deposit_event(Event::<T>::NativeTokenBurned(who, burned_amt.saturated_into()));
 			Ok(())
 		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(<T as Config>::WeightInfo::parachain_withdraw(1))] // TODO: @zktony Benchmarking
+		pub fn ethereum_withdraw(
```
Comment (serhii-temchenko) [inline pallets/thea-executor/src/lib.rs:383]: No tests for this extrinsic

File: pallets/thea-executor/src/lib.rs [hunk 2]
```diff
@@ -402,6 +403,30 @@ pub mod pallet {
...
+			origin: OriginFor<T>,
+			asset_id: u128,
+			amount: u128,
+			beneficiary: H160,
+			pay_for_remaining: bool,
+			pay_with_tokens: bool,
+		) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-executor/src/lib.rs]: Would be nice to have docblock here

File: pallets/thea/src/lib.rs [hunk 3]
```diff
@@ -86,31 +87,57 @@ pub trait TheaWeightInfo {
 	fn remove_thea_network() -> Weight;
 }
 
-#[frame_support::pallet]
+#[frame_support::pallet(dev_mode)]
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Should this `dev_mode` be removed?

File: pallets/thea/src/lib.rs [hunk 4]
```diff
@@ -205,31 +336,75 @@ pub mod pallet {
...
-			T::Executor::execute_deposits(payload.network, payload.data.clone());
-			<IncomingNonce<T>>::insert(payload.network, payload.nonce);
-			Self::deposit_event(Event::<T>::TheaPayloadProcessed(payload.network, payload.nonce));
-			// Save the incoming message for some time
-			<IncomingMessages<T>>::insert(payload.network, payload.nonce, payload);
+			let signer = ensure_signed(origin)?;
+			let config = <NetworkConfig<T>>::get(payload.network);
+
+			if stake < config.min_stake {
+				return Err(Error::<T>::NotEnoughStake.into())
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:364]: Should be tested this condition

File: pallets/thea/src/lib.rs [hunk 5]
```diff
@@ -296,6 +481,140 @@ pub mod pallet {
 			});
 			Ok(())
 		}
+
+		/// Signed outgoing messages
+		#[pallet::call_index(6)]
+		#[pallet::weight(10_000)] // TODO: @zktony Benchmarking.
+		pub fn submit_signed_outgoing_messages(
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:491]: It should be tested

File: pallets/thea/src/lib.rs [hunk 6]
```diff
@@ -296,6 +481,140 @@ pub mod pallet {
...
+			}
+
+			Ok(())
+		}
+
+		/// Report misbehaviour as fisherman
+		#[pallet::call_index(7)]
+		#[pallet::weight(10_000)] // TODO: @zktony Benchmarking
+		#[transactional]
+		pub fn report_misbehaviour(
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:527]: Should be tested

File: pallets/thea/src/lib.rs [hunk 7]
```diff
@@ -296,6 +481,140 @@ pub mod pallet {
...
+				},
+			}
+			Ok(())
+		}
+
+		/// Handle misbehaviour via governance
+		#[pallet::call_index(8)]
+		#[pallet::weight(10_000)] // TODO: @zktony Benchmarking
+		#[transactional]
+		pub fn handle_misbehaviour(
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:569]: Should be tested

File: primitives/thea/src/types.rs
Comment (serhii-temchenko) [inline primitives/thea/src/types.rs:1]: Please add documentation for each new member (pub api) of this file

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/thea-executor/src/lib.rs (hunk 2 above)
Comment (zktony) [reply pallets/thea-executor/src/lib.rs]: done

File: pallets/thea-executor/src/lib.rs (hunk 1 above)
Comment (zktony) [reply pallets/thea-executor/src/lib.rs:383]: Added one test case to test the happy path. Parachain_withdraw tests already cover other cases.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -115,7 +115,7 @@ jobs:
       - name: Check Clippy
         run: cargo clippy -- -D warnings
       - name: Test Project
-        run: RUSTFLAGS="-D warnings" cargo test --workspace
+        run: cargo test --workspace
```
Comment (serhii-temchenko) [inline .github/workflows/ci.yml]: Can this be moved back?

File: pallets/thea-executor/src/mock.rs [hunk 9]
```diff
@@ -129,6 +129,8 @@ impl pallet_assets::Config for Test {
 	type Extra = ();
 	type CallbackHandle = ();
 	type WeightInfo = ();
+	// #[cfg(feature = "runtime-benchmarks")]
+	// type BenchmarkHelper = AssetU128;
```
Comment (serhii-temchenko) [inline pallets/thea-executor/src/mock.rs]: Please remove

File: pallets/thea-message-handler/src/lib.rs [hunk 10]
```diff
@@ -181,44 +187,50 @@ pub mod pallet {
 
 		/// Handles the verified incoming message
 		#[pallet::call_index(1)]
-		#[pallet::weight(<T as Config>::WeightInfo::incoming_message())]
+		#[pallet::weight(<T as Config>::WeightInfo::incoming_message())] // TODO: @zktony Benchmark again for updated logic
```
Comment (serhii-temchenko) [inline pallets/thea-message-handler/src/lib.rs]: I assume this was updated? Can we remove todo?

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

**Comment by Gauthamastro** [comment]
@zktony , The pending tests.

- [x] on_init in thea
- [x] paramaterization of on_initialize function in thea 
- [x] validate_signed_outgoing_message
- [x] submit_incoming_message
- [x] match statement inside incoming_message extrinsic in thea message handler.
- [x] validate_incoming_message in thea message handler ( else blocks are not tested )
- [x] on_initialize  in thea executor
- [x] paramaterization of on_initialize function in thea executor
- [x] add_signature() else condition, contains_signature() methods of SignedMessage in types
- [x] amount_in_foreign_decimals() method of Deposit type in types.rs

**Review by serhii-temchenko — APPROVED** [review]

**Review by zktony — APPROVED** [review]

**Pull request merged**